src/rgw/rgw_lc.cc

```diff
@@ -1923,13 +1970,94 @@ time_t RGWLC::thread_stop_at()
   return time(nullptr) + interval;
 }
 
+int RGWLC::process_bucket(int index, int max_lock_secs, LCWorker* worker,
+			  const std::string& bucket_entry_marker,
+			  bool once = false)
+{
+  ldpp_dout(this, 5) << "RGWLC::process_bucket(): ENTER: "
+	  << "index: " << index << " worker ix: " << worker->ix
+	  << dendl;
+
+  int ret = 0;
+  rgw::sal::LCSerializer* lock = sal_lc->get_serializer(lc_index_lock_name,
+							obj_names[index],
+							std::string());
+  //string = bucket_name:bucket_id, start_time, int = LC_BUCKET_STATUS
+  rgw::sal::Lifecycle::LCEntry entry;
+  if (max_lock_secs <= 0)
+    return -EAGAIN;
+
+  utime_t time(max_lock_secs, 0);
+retry_lock:
+  ret = lock->try_lock(this, time, null_yield);
+  if (ret == -EBUSY || ret == -EEXIST) {
+    /* already locked by another lc processor */
+    ldpp_dout(this, 0) << "RGWLC::process() failed to acquire lock on "
+		       << obj_names[index] << ", sleep 5, try again" << dendl;
+    sleep(5);
+    goto retry_lock; // XXXX really retry forever?
+  }
+  if (ret < 0)
+    return 0;
+
+  if (! (cct->_conf->rgw_lc_lock_max_time == 9969)) {
+    ret = sal_lc->get_entry(obj_names[index], bucket_entry_marker, entry);
+    if (ret >= 0) {
+      if (entry.status == lc_processing) {
+	if (expired_session(entry.start_time)) {
+	  ldpp_dout(this, 5) << "RGWLC::process_bucket(): STALE lc session found for: " << entry
+			     << " index: " << index << " worker ix: " << worker->ix
+			     << " (clearing)"
+			     << dendl;
+	} else {
+	  ldpp_dout(this, 5) << "RGWLC::process_bucket(): ACTIVE entry: "
+			     << entry
+			     << " index: " << index
+			     << " worker ix: " << worker->ix
+			     << dendl;
+	  goto exit;
```

Comment:
lots of `goto exit`s in this function that don't unlock or free the lock. would very much prefer to use RAII (unique_ptr, unique_lock) instead of goto for this

Reply:
I fixed the error returns while testing, and repushed.  I'm happy to use raii, if the serializer already supports it

Reply:
after the call to `lock->try_lock()` succeeds, you can 'adopt' it with a scoped_lock:
```c++
auto unlock = std::scoped_lock{std::adopt_lock, *lock};
```

Reply:
I already use scope_guard else where, can do that